File: backend-go/internal/domain/topology.go
package domain

// ResourceType identifies the kind of infrastructure resource
type ResourceType string

const (
	ResourcePod        ResourceType = "pod"
	ResourceService    ResourceType = "service"
	ResourceDeployment ResourceType = "deployment"
	ResourceNode       ResourceType = "node"
	ResourceNamespace  ResourceType = "namespace"
	ResourceEC2        ResourceType = "ec2"
	ResourceRDS        ResourceType = "rds"
	ResourceVPC        ResourceType = "vpc"
	ResourceSubnet     ResourceType = "subnet"
)

// HealthStatus describes the health of a resource
type HealthStatus string

const (
	HealthHealthy   HealthStatus = "healthy"
	HealthDegraded  HealthStatus = "degraded"
	HealthUnhealthy HealthStatus = "unhealthy"
	HealthUnknown   HealthStatus = "unknown"
)

// TopologyNode represents a single infrastructure resource
type TopologyNode struct {
	ID           string            `json:"id"`
	Name         string            `json:"name"`
	ResourceType ResourceType      `json:"resource_type"`
	Namespace    *string           `json:"namespace,omitempty"`
	Labels       map[string]string `json:"labels,omitempty"`
	Health       HealthStatus      `json:"health"`
	Metadata     map[string]any    `json:"metadata,omitempty"`
}

// TopologyEdge represents a relationship between two resources
type TopologyEdge struct {
	Source   string         `json:"source"`
	Target   string         `json:"target"`
	Relation string         `json:"relation"`
	Metadata map[string]any `json:"metadata,omitempty"`
}

// InfraTopology is the full resource graph
type InfraTopology struct {
	Nodes     []TopologyNode `json:"nodes"`
	Edges     []TopologyEdge `json:"edges"`
	Timestamp *string        `json:"timestamp,omitempty"`
}

// ResilienceScore summarizes system resilience
type ResilienceScore struct {
	Overall         float64            `json:"overall"`
	Categories      map[string]float64 `json:"categories,omitempty"`
	Recommendations []string           `json:"recommendations,omitempty"`
	Details         *string            `json:"details,omitempty"`
}

File: backend/models/topology.py
from enum import Enum
from typing import Optional
from pydantic import BaseModel, Field


class ResourceType(str, Enum):
    POD = "pod"
    SERVICE = "service"
    DEPLOYMENT = "deployment"
    NODE = "node"
    NAMESPACE = "namespace"
    EC2 = "ec2"
    RDS = "rds"
    VPC = "vpc"
    SUBNET = "subnet"


class HealthStatus(str, Enum):
    HEALTHY = "healthy"
    DEGRADED = "degraded"
    UNHEALTHY = "unhealthy"
    UNKNOWN = "unknown"


class TopologyNode(BaseModel):
    id: str
    name: str
    resource_type: ResourceType
    namespace: Optional[str] = None
    labels: dict[str, str] = Field(default_factory=dict)
    health: HealthStatus = HealthStatus.UNKNOWN
    metadata: dict = Field(default_factory=dict)


class TopologyEdge(BaseModel):
    source: str
    target: str
    relation: str = "connects_to"
    metadata: dict = Field(default_factory=dict)


class InfraTopology(BaseModel):
    nodes: list[TopologyNode] = Field(default_factory=list)
    edges: list[TopologyEdge] = Field(default_factory=list)
    timestamp: Optional[str] = None


class ResilienceScore(BaseModel):
    overall: float = Field(ge=0.0, le=100.0)
    categories: dict[str, float] = Field(default_factory=dict)
    recommendations: list[str] = Field(default_factory=list)
    details: Optional[str] = None
